src/aatoolbox/datasources/glofas/forecast.py

```diff
@@ -0,0 +1,266 @@
+
+import xarray as xr
+from dateutil import rrule
+
+from aatoolbox.config.countryconfig import CountryConfig
+from aatoolbox.datasources.glofas import glofas
+from aatoolbox.utils.check_file_existence import check_file_existence
+from aatoolbox.utils.geoboundingbox import GeoBoundingBox
+
+logger = logging.getLogger(__name__)
+
+
+class _GlofasForecastBase(glofas.Glofas):
+    """Base class for all GloFAS forecast data downloading and processing."""
+
+    def __init__(
+        self,
+        country_config: CountryConfig,
+        geo_bounding_box: GeoBoundingBox,
+        leadtime_max: int,
+        start_date: datetime,
+        end_date: datetime,
+        cds_name: str,
+        system_version: str,
+        product_type: Union[str, List[str]],
+        date_variable_prefix: str,
+        frequency: int,
+        coord_names: List[str],
+    ):
+        super().__init__(
+            country_config=country_config,
+            geo_bounding_box=geo_bounding_box,
+            start_date=start_date,
+            end_date=end_date,
+            cds_name=cds_name,
+            system_version=system_version,
+            product_type=product_type,
+            date_variable_prefix=date_variable_prefix,
+            frequency=frequency,
+            coord_names=coord_names,
+            leadtime_max=leadtime_max,
+        )
+
+    @check_file_existence
+    def _load_single_file(
+        self, input_filepath: Path, filepath: Path, clobber: bool
+    ) -> xr.Dataset:
+        return _read_in_ensemble_and_perturbed_datasets(
+            filepath=input_filepath
+        )
+
+
+def _read_in_ensemble_and_perturbed_datasets(filepath: Path):
+    """Read in forecast and reforecast data.
+
+    The GloFAS forecast and reforecast data GRIB files contain two
+    separate datasets: the control member, generated from the most accurate
+    estimate of current conditions, and the perturbed forecast, which
+    contains N ensemble members created by perturbing the control forecast.
+
+    This function reads in both datasets and creates an N+1 (perturbed
+    + control) ensemble.
+    See `this paper <https://hess.copernicus.org/preprints/hess-2020-532/>`_
+    for more details.
+    """
+    ds_list = []
+    for data_type in ["cf", "pf"]:
+        ds = xr.load_dataset(
+            filepath,
+            engine="cfgrib",
+            backend_kwargs={
+                "indexpath": "",
+                "filter_by_keys": {"dataType": data_type},
+            },
+        )
+        # Extra processing require for control forecast
+        if data_type == "cf":
+            ds = ds.expand_dims(dim="number")
+        ds_list.append(ds)
+    return xr.combine_by_coords(ds_list, combine_attrs="drop_conflicts")
+
+
+class GlofasForecast(_GlofasForecastBase):
+    """
+    Class for downloading and processing GloFAS forecast data.
+
+    The GloFAS forecast dataset is a global raster presenting river
+    discharnge forecast from 26 May, 2021 until present day, see
```

Comment:
discharge

Reply:
Is it really just from 26 May 2021? Paper linked mentions 1 January 2020 and then 20 years of reforests, so should it be from 2020?

Reply:
It's a bit complicated. I mentioned a bit in the docs, but the version of the forecast from January 2020 was only the pre-release and some bugs were ironed out before the final release on 26 May 2021. This does leave an unfortunate and recent gap, I can get in touch with the GLoFAS team to understand how bad it would be to use the pre-release too, it would a few tweaks to get that supported. In the meantime I will make the docs more clear.